File: src/language/Language.php
<?php

namespace pocketcloud\cloud\language;

use pocketcloud\cloud\config\impl\MainConfig;
use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\util\FileUtils;
use pocketcloud\cloud\util\trait\RegistryTrait;
use pocketcloud\cloud\util\Utils;
use RuntimeException;
use Throwable;
use const pocketcloud\CLOUD_PATH;
use const pocketcloud\IN_GAME_PATH;

/**
 * @method static Language GERMAN()
 * @method static Language ENGLISH()
 */
final class Language {
    use RegistryTrait;

    public const string FALLBACK = "en";

    public static function init(): void {
        self::register("german", new Language(
            "German",
            IN_GAME_PATH . "de_DE.yml",
            ["de_DE", "ger", "Deutsch"],
            DefaultMessages::MESSAGES_DE
        ));

        self::register("english", new Language(
            "English",
            IN_GAME_PATH . "en_US.yml",
            ["en_US", "en", "Englisch"],
            DefaultMessages::MESSAGES
        ));
    }

    public static function current(): Language {
        return self::get(MainConfig::getInstance()->getLanguage());
    }

    public static function fallback(): Language {
        return self::get(self::FALLBACK);
    }

    public static function get(string $name): ?Language {
        /** @var Language $language */
        return array_find(self::getAll(), fn($language) => $language->getName() == $name || in_array($name, $language->getAliases()));
    }

    /** @var array<string, string> */
    private array $messages;

    public function __construct(
        private readonly string $name,
        private readonly string $filePath,
        private readonly array $aliases,
        array $defaultMessages = []
    ) {
        if (file_exists($this->filePath)) {
            try {
                $messages = FileUtils::parseYamlFile($this->filePath);
                if ($messages === null) throw new RuntimeException("Failed to parse messages from " . str_replace(CLOUD_PATH, "", $this->filePath));
                Utils::fillMissingKeys($messages, $defaultMessages, $affectedKeys);
                if ($affectedKeys > 0) {
                    CloudLogger::get()->info("Incomplete language file found: §b{}§r, completed the file with the missing lang keys.", str_replace(CLOUD_PATH, "", $this->filePath));
                    FileUtils::emitYamlFile($this->filePath, $messages, YAML_UTF8_ENCODING);
                }

                $this->messages = $messages;
            } catch (Throwable $exception) {
                $this->messages = $defaultMessages;
                CloudLogger::get()->exception($exception);
            }
        } else {
            CloudLogger::get()->info("Language file not found: §b{}§r, generating...", str_replace(CLOUD_PATH, "", $this->filePath));
            $this->messages = $defaultMessages;
            FileUtils::emitYamlFile($this->filePath, $this->messages, YAML_UTF8_ENCODING);
        }
    }

    public function translate(string $key, array $args = []): string {
        $message = str_replace("{PREFIX}", $this->messages["inGame.prefix"] ?? "", $this->messages[$key] ?? $key);
        foreach ($args as $k => $arg) $message = str_replace("%" . $k . "%", $arg, $message);
        return $message;
    }

    public function getName(): string {
        return $this->name;
    }

    public function getAliases(): array {
        return $this->aliases;
    }

    public function getMessages(): array {
        return $this->messages;
    }
}

File: src/main/java/de/pocketcloud/cloud/language/Language.java
package de.pocketcloud.cloud.language;

import de.pocketcloud.cloud.config.impl.MainConfig;
import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.util.ArrayUtils;
import de.pocketcloud.cloud.util.FileUtils;
import de.pocketcloud.cloud.util.PocketCloudPaths;
import lombok.Getter;
import lombok.experimental.Accessors;
import org.jetbrains.annotations.Nullable;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

@Getter
@Accessors(fluent = true)
public enum Language {

    ENGLISH(PocketCloudPaths.storage().inGame().with("de_DE.yml").asPath(), new String[]{"de_DE", "ger", "Deutsch"}, DefaultMessages.MESSAGES),
    GERMAN(PocketCloudPaths.storage().inGame().with("en_US.yml").asPath(), new String[]{"en_US", "en", "English"}, DefaultMessages.MESSAGES_DE);

    private final Map<String, String> messages;

    private final Path filePath;
    private final String[] aliases;
    private final Map<String, String> defaultMessages;

    Language(Path filePath, String[] aliases, Map<String, String> defaultMessages) {
        this.filePath = filePath;
        this.aliases = aliases;
        this.defaultMessages = defaultMessages;

        if (Files.isRegularFile(filePath)) {
            Map<String, Object> messages = FileUtils.parseYamlFile(filePath);
            Map<String, String> realMessages = new HashMap<>();
            for (Map.Entry<String, Object> entry : messages.entrySet()) {
                if (entry.getValue() instanceof String) {
                    realMessages.put(entry.getKey(), (String) entry.getValue());
                }
            }

            AtomicInteger affected = new AtomicInteger();
            ArrayUtils.fillMissingKeys(realMessages, defaultMessages, affected, true);
            if (affected.get() > 0) {
                CloudLogger.get().info("Incomplete language file §b{}§r, completing the file with the missing language keys...", filePath.toString());
                FileUtils.emitYamlFile(filePath, realMessages);
            }

            this.messages = realMessages;
        } else {
            CloudLogger.get().info("Language file §b{} §rnot found, generating...", filePath.toString());
            this.messages = defaultMessages;
            FileUtils.emitYamlFile(filePath, defaultMessages);
        }
    }

    public String translate(String message, Object ... args) {
        message = message.replace("{PREFIX}", this.messages.getOrDefault("inGame.prefix", ""));
        for (int i = 0; i < args.length; i++) {
            message = message.replace("{" + i + "}", args[i].toString());
        }
        return message;
    }

    public static Language current() {
        try {
            return Language.valueOf(MainConfig.instance().getLanguage());
        } catch (IllegalArgumentException e) {
            return fallback();
        }
    }

    public static Language fallback() {
        return Language.ENGLISH;
    }

    @Nullable
    public static Language get(String name) {
        return Arrays.stream(values()).filter(l -> l.name().equalsIgnoreCase(name) || Arrays.stream(l.aliases()).anyMatch(a -> a.equalsIgnoreCase(name)))
                .findFirst()
                .orElse(null);
    }
}